db_test.go

```diff
@@ -0,0 +1,236 @@
+	Equals(t, longestDowntimeEnd.Format(timeFormat), stat.longestDowntime.end.Format(timeFormat))
+
+}
+
+func TestSaveHostname(t *testing.T) {
+	// There are many fields, so many things could go wrong; that's why this elaborate test.
+	arg := []string{"localhost", "8001"}
+	s := newDb(arg, ":memory:")
+	defer s.db.Close()
+	stat := mockStats()
+
+	err := s.saveHostNameChange(stat.hostnameChanges)
+	Nil(t, err)
+	// testing the host names if they are properly written
+	query := `SELECT
+	hostname_changed_to, hostname_change_time
+	FROM ` + fmt.Sprintf("%s WHERE event_type IS '%s';", s.tableName, eventTypeHostnameChange)
+
+	rows, err := s.db.Query(query)
+	Nil(t, err)
+
+	idx := 0
+	for rows.Next() {
+		var hostName string
+		var cTime time.Time
+		err = rows.Scan(&hostName, &cTime)
+		Nil(t, err)
+
+		actualHost := stat.hostnameChanges[idx]
+		idx++
+		Equals(t, hostName, actualHost.Addr.String())
+		Equals(t, cTime.Format(timeFormat), actualHost.When.Format(timeFormat))
+	}
+	Equals(t, idx, len(stat.hostnameChanges))
+	rows.Close()
+}
+
+func hostNameChange() []hostnameChange {
+	ipAddresses := []string{
+		"192.168.1.1",
+		"10.0.0.1",
+		"172.16.0.1",
+		"2001:0db8:85a3:0000:0000:8a2e:0370:7334",
+	}
+	var hostNames []hostnameChange
+	for i, ip := range ipAddresses {
+		host := hostnameChange{
+			Addr: netip.MustParseAddr(ip),
+			When: time.Now().Add(time.Duration(i) * time.Minute),
+		}
+		hostNames = append(hostNames, host)
+	}
+	return hostNames
+}
+
+func mockStats() stats {
+	stat := stats{
+		startTime:           time.Now(),
+		endTime:             time.Now().Add(10 * time.Minute),
+		lastSuccessfulProbe: time.Now().Add(1 * time.Minute),
+		// lastUnsuccessfulProbe:  time.Now().Add(2 * time.Minute),
+		retriedHostnameLookups: 10,
+		longestUptime: longestTime{
+			start:    time.Now().Add(20 * time.Second),
+			end:      time.Now().Add(80 * time.Second),
+			duration: time.Minute,
+		},
+		longestDowntime: longestTime{
+			start:    time.Now().Add(20 * time.Second),
+			end:      time.Now().Add(140 * time.Second),
+			duration: time.Minute * 2,
+		},
+		userInput: userInput{
+			ip:       netip.MustParseAddr("192.168.1.1"),
+			hostname: "example.com",
+			port:     1234,
+		},
+		totalUptime:             time.Second * 32,
+		totalDowntime:           time.Second * 60,
+		totalSuccessfulProbes:   201,
+		totalUnsuccessfulProbes: 123,
+		rttResults: rttResult{
+			min:     2.832,
+			average: 3.8123,
+			max:     4.0932,
+		},
+
+		hostnameChanges: hostNameChange(),
+	}
+
+	return stat
+}
+
+// Equals compares two values.
+// This is for avoiding code duplications.
+func Equals[T comparable](t *testing.T, value, want T) {
+	t.Helper()
+	if want != value {
+		t.Errorf("wanted %v; got %v", want, value)
+		t.FailNow()
+	}
+}
+
+// Nil compares a value to nil, in some cases you may need to do `Equals(t, value, nil)`
+func Nil(t *testing.T, value any) {
```

Comment:
I would use a different name instead of `Nil`. This is a bit confusing at the first glance. Even my editor shows it in a different color.